Remaining CWG updates from Wrocław. (#207)
brevzin/cpp_proposals: 2996_reflection/reflection.md

  * [[7.8-1]{.pnum} If both constructs are within the `$member-specification$` of a class `C` ([class.mem.general]{.sref}) and one such construct (call it `$X$`) is in a complete-class context of `C` but the other (call it `$Y$`) is not, then `$X$` semantically follows `$Y$`.]{.addu}
  * [[7-8.#]{.pnum} Otherwise, if the syntactic endpoint of the tokens that describe one such construct (call it `$X$`) occurs lexically after the syntactic endpoint of the tokens that describe the other (call it `$Y$`), then `$X$` semantically follows `$Y$`.]{.addu}

>   [During the analysis and translation of tokens and instantiations, certain expressions are evaluated ([expr.const]{.sref}). Diagnosable rules ([intro.compliance.general]{.sref}) that apply to constructs that semantically follow a plainly constant-evaluated expression `$P$` are considered in a context where `$P$` has been evaluated exactly once.]{.addu}

Reviewer: Don't add srefs in wording, it's just noisy.

Author: Fiiiine

```
:::

### [basic.pre]{.sref} Preamble {-}

Add type aliases and namespace aliases to the list of entities in paragraph 3. As drive-by fixes, remove "variable", "object", "reference", and "template specialization"; replace "class member" with "non-static data member", since all other cases are subsumed by existing one. Add "template parameters" and "`$init-capture$`s", which collectively subsume "packs".

::: std

> [3]{.pnum} An _entity_ is a [value, object, reference]{.rm} [variable,]{.addu} structured binding, function, enumerator, type, [type alias]{.addu}, [class]{.rm} [non-static data]{.addu} member, bit-field, template, template specialization, namespace, [namespace alias, template parameter]{.addu}, or [`$init-capture$`]{.addu} [pack]{.rm}.

Reviewer: I feel like we should have a note for other things that are entities as a follow on from here. Like a function parameter pack is an entity because it is a pack of variables or something?

Author: I believe that Davis's Grand Design wants us to move towards "entity" being roughly "eternal thing whose declarations are found by name lookup" - a pack can be found by name lookup, ergo entity.

:::

Introduce a notion of an "underlying entity" in paragraph 5, and utilize it for the definition of a name "denoting" an entity. Type aliases are now entities, so also modify accordingly.

::: std
[5]{.pnum} Every name is introduced by a _declaration_, which is a

* [#.#]{.pnum} `$name-declaration$`, `$block-declaration$`, or `$member-declaration$` ([dcl.pre]{.sref}, [class.mem]{.sref}),

[...]

* [#.14]{.pnum} implicit declaration of an injected-class-name ([class.pre]{.sref}).

> [The _underlying entity_ of an entity is that entity unless otherwise specified. A name _denotes_ the underlying entity of the entity declared by each declaration that introduces the name. [An entity `$E$` is denoted by the name (if any) that is introduced by a declaration of `$E$` or by a `$typedef-name$` introduced by a declaration specifying `$E$`.]{.rm}]{.addu}

Reviewer: The `{.rm}` and the `{.addu}` should be separate, it's weird to see it nested.

Author: Oops, didn't mean to do that. Thanks, fixed.

> [[Entities whose underlying entity is not themselves include type aliases and namespace aliases.]{.note}]{.addu}

Reviewer: I'd phrase this as like "A type alias and a namespace alias are examples of entities whose underlying entity is distinct from itself."? Something like that.

:::

Modify paragraph 6 such that denoting a variable by its name finds the variable, not the associated object.

::: std
[6]{.pnum} A _variable_ is introduced by the declaration of a reference other than a non-static data member or of an object. [The variable's name, if any, denotes the reference or object.]{.rm}
:::

### [basic.def]{.sref} Declarations and definitions {-}

> Modify the third sentence of paragraph 1 to clarify that type aliasesare now entities.

Reviewer: ```suggestion
Modify the third sentence of paragraph 1 to clarify that type aliases are now entities.
```

Author: Done.

::: std
[1]{.pnum} [...] A declaration of an entity [or `$typedef-name$`]{.rm} `$X$` is a redeclaration of `$X$` if another declaration of `$X$` is reachable from it ([module.reach]{.sref}). [...]

:::

Since namespace aliases are now entities, but their declarations are not definitions, add `$namespace-alias-definition$` to the list of declarations in paragraph 2, just before `$using-declaration$`:

::: std
[2]{.pnum} Each entity declared by a `$declaration$` is also _defined_ by that declaration unless:

* [#.#]{.pnum} it declares a function without specifying the function's body ([dcl.fct.def]{.sref}),

[...]

* [2.10]{.pnum} it is an `$alias-declaration$` ([dcl.typedef]{.sref}),

> * [[2.11-]{.pnum} it is `$namespace-alias-definition$` ([namespace.alias]{.sref}),]{.addu}

Reviewer: ```suggestion
* [[2.11-]{.pnum} it is a `$namespace-alias-definition$` ([namespace.alias]{.sref}),]{.addu}
```

Author: Done.

the program is ill-formed; a diagnostic is required only if the definable item is attached to a named module and a prior definition is reachable at the point where a later definition occurs. [...]
:::

Prefer the verb "denote" in bullet 15.5 to emphasize that ODR "looks through" aliases, and clarify that objects are not entities in bullet 15.5.2.

::: std

> - [15.5]{.pnum} In each such definition, corresponding names, looked up according to [basic.lookup]{.sref}, shall [refer to]{.rm} [denote]{.addu} the same entity, after overload resolution ([over.match]{.sref}) and after matching of partial template specialization ([temp.over]{.sref}), except that a name can refer to

Reviewer: I don't know how else to say this, but this paragraph sounds like William Shatner. There are too many commas and I think we're missing some words. And apparently that's just status quo, so... alrighty then.

[13]{.pnum} A declaration `$D$` _names_ an entity `$E$` if

* [13.1]{.pnum} `$D$` contains a `$lambda-expression$` whose closure type is `$E$`,

> * [13.1+]{.pnum} [`$D$` contains an expression of type `std::meta::info` that represents an entity that is either `$E$` or whose underlying entity is `$E$`,]{.addu}

Reviewer: So it's intentional that `using T = C;` names both `T` and `C`? Just want to check.

Author: On second thought - No, it only needs to name `T` (thanks for the sanity check).

The notion of when a declaration _names_ an entity is mostly used to define when a declaration is an exposure (i.e., when it names a TU-local entity). But since any type alias or namespace alias is already TU-local, any declaration that names the alias is already an exposure - so no need to really pay any attention to the underlying entity.

Note - This section needs some more attention and scrutiny. CWG asked for examples of what I'm trying to achieve here in Poland; I still need to get around to that. I'll probably turn to that after this PR is merged.

template <auto V> constexpr int e = [:V:];   // splice-expression
constexpr int f = template [:^^e:]<^^S::a>;  // splice-expression

> auto g = typename [:^^int:](42);  // [:^^int:] is not an expression

Reviewer: Is this... supposed to be an error? I don't know what this comment means.

Author: Not an error - Let me know if this is clearer.

[The implicit transformation ([expr.prim.id]{.sref}) whereby an `$id-expression$` denoting a non-static member becomes a class member access does not apply to a `$splice-expression$`.]{.note}

> [#]{.pnum} For a `$splice-expression$` that designates a function or function template, overload resolution is performed ([over.match]{.sref}, [temp.over]{.sref}) from an initial set of candidate functions containing only the entity designated by the `$splice-expression$`. The best viable function selected by overload resolution is _designated in a manner exempt from access rules_.

Reviewer: Stray `_` in front of designated?

Author: Nope, "designated in a manner exempt from access rules" is the term being introduced.

Reviewer: Ah just couldn't see the other `_`

[6]{.pnum} If `E2` [is]{.rm} [designates]{.addu} a bit-field, `E1.E2` is a bit-field. [...]

> [7]{.pnum} If `E2` [designates an entity]{.addu} [is]{.rm} declared to have type "reference to `T`", then `E1.E2` is an lvalue of type `T`. [If]{.rm} [In that case, if]{.addu} `E2` [is]{.rm} [designates]{.addu} a static data member, `E1.E2` designates the object or function to which the reference is bound, otherwise `E1.E2` designates the object or function to which the corresponding reference member of `E1` is bound. Otherwise, one of the following rules applies.

Reviewer: Maybe: If `E2` designates an entity that is declared to have [...]?

Author: I dig it.

[#]{.pnum} A `$reflect-expression$` of the form `^^ ::` represents the global namespace. A `$reflect-expression$` of the form `^^ $qualified-namespace-specifier$`
is interpreted as follows:

- [#]{.pnum} If lookup of the `$qualified-namespace-specifier$` unambiguously finds a `$namespace-alias-definition$`, the `$reflect-expression$` represents the namespace alias declared by that `$namespace-alias-definition$`.
- [#]{.pnum} Otherwise, the `$reflect-expression$` represents the namespace denoted by the `$qualified-namespace-specifier$`.

[#]{.pnum} A `$reflect-expression$` of the form `^^ $nested-name-specifier$@~_opt_~@ $template-name$` represents the primary class template, function template, primary variable template, or alias template denoted by `$template-name$`.

[#]{.pnum} A `$reflect-expression$` of the form `^^ $nested-name-specifier$@~_opt_~@ $concept-name$` represents the concept denoted by `$concept-name$`.

[#]{.pnum} A `$reflect-expression$` of the form `^^ $type-id$` is interpreted as follows:

> * [#.#]{.pnum} If `$type-id$` is a `$typedef-name$` that was introduced by the declaration of a template parameter, the `$reflect-expression$` represents the type denoted by that type alias.

Reviewer: Should this be "If _the_ `$type-id$`"? Throughout?

Author: Yep, think so. Done.

```
:::

> [#]{.pnum} A `$reflect-expression$` that could be validly interpreted as `^^ $template-name$` is never interpreted as `^^ $id-expression$`, and is interpreted as `^^ $type-id$` if and only if the operand refers to the current instantiation ([temp.dep.type]{.sref}).

Author: I like them, but fine. Done.

[5+]{.pnum} If both operands are of type `std::meta::info`, comparison is defined as follows:

* [5+.#]{.pnum} If one operand is a null reflection value, then they compare equal if and only if the other operand is also a null reflection value.
* [5+.#]{.pnum} Otherwise, if one operand represents a value, then they compare equal if and only if the other operand represents a value that is template-argument-equivalent ([temp.type]{.sref}).
* [5+.#]{.pnum} Otherwise, if one operand represents an object, then they compare equal if and only if the other operand represents the same object.
* [5+.#]{.pnum} Otherwise, if one operand represents an entity, then they compare equal if and only if the other operand represents the same entity.

> * [5+.#]{.pnum} Otherwise, if one operand represents a direct base class relationship, then they compare equal if and only if the other operand represents the same relationship to the same direct base class.

Reviewer: Why not just "[...] represents the same direct base class relationship"?

Author: Guess I was feeling prose-y. Done.

> * [5+.#]{.pnum} Otherwise, both operands `O@~_1_~@` and `O@~_2_~@` represent data member descriptions. The operands compare equal if and only if the data member descriptions represented by `O@~_1_~@` and `O@~_2_~@` compare equal ([class.mem.general]{.sref}).

Reviewer: We do not know at this point that both represent data member descriptions. We don't have a bullet for "same kind" - so we could be that one is a data member description and the other is a type.

Author: We do know that: We have a bullet for "if _one_ operand represents an entity". So a comparison between a data member description and a type would fall under that bullet.

Reviewer: Ok, data member description vs value then.

Author: Same thing - All of the other bullets are phrased the same ways:
- If **one** operand is a null reflection, ...
- Otherwise, if **one** operand represents a value, ...
- Otherwise, if **one** operand represents an object, ...
- Otherwise, if **one** operand represents an entity, ...
- Otherwise, if **one** operand represents a direct base class relationship...

At the end, we have "data member description", and if _both_ operands failed to fall into _any_ of the other buckets, then it must be that both of them fall into this one.

Reviewer: Ah. Got it. Yep, you're right.

  // error: the invocation of make_decl(R) in the requires clause produces an
  // injected declaration but is not plainly constant-evaluated

template <typename> struct T {
  static constexpr bool b4 = make_decl(4);
    // error: target scope of the injected declaration is not
    // instantatiation-sequenced with the expression
}

struct S;
consteval bool complete_decl();

>   // produces an injected declaration completing S

Reviewer: Can you just make this do:
```cpp
consteval bool complete_decl() {
  define_aggregate(^^S{}, {});
  return true;
}
```
Makes the example easier to follow.

constexpr auto v = (complete_decl(), 1) + sizeof(^^S);

>   // ill-formed, no diagnostic required

Reviewer: Don't need the `(x, 1)`, can just `complete_decl() + sizeof(^^S)`.

Author: Done.

:::

Add a new paragraph to the end of the section defining _data member description_:

::: std
::: addu

> [29+]{.pnum} A _data member description_ is a quintuple (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) describing the potential declaration of a nonstatic data member where

Reviewer: ```suggestion
[29+]{.pnum} A _data member description_ is a tuple (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) describing the potential declaration of a non-static data member where
```

Author: Daveed pointed out that "quintuple" has precedence in the Standard:
https://eel.is/c++draft/over.built#9

Maybe we should keep as-is?

- [29+.#]{.pnum} `$T$` is a type or type alias,
- [29+.#]{.pnum} `$N$` is an `$identifier$` or `-1`,
- [29+.#]{.pnum} `$A$` is an alignment or `-1`,
- [29+.#]{.pnum} `$W$` is a bit-field width or `-1`, and

> - [29+.#]{.pnum} `$NU$` is a boolean value that is `true` if the potentially-declared member would be declared with the attribute `no_unique_address` and `false` otherwise.

Reviewer: I think either call this `$U$` (as in unique) or full `$NUA$`. 

Also why does this bullet get a full description but the three previous ones do not?

Author: Went with `$NUA$`, since I find it confusing that `U` would mean "non-unique".

Good point re: full description - I've added a note with a more full explanation and with cross-ref to `data_member_spec` and to `define_aggregate`.

> Two data member descriptions are equal if each of their respective components are same types, same identifiers, or equal values.

Reviewer: ```suggestion
Two data member descriptions are equal if each of their respective components are same types, same identifiers, and equal values.
```

Author: Done.

::: std
::: addu
[29+]{.pnum} A _data member description_ is a quintuple (`$T$`, `$N$`, `$A$`, `$W$`, `$NUA$`) describing the potential declaration of a nonstatic data member where

- [29+.#]{.pnum} `$T$` is a type or type alias,
- [29+.#]{.pnum} `$N$` is an `$identifier$` or `-1`,
- [29+.#]{.pnum} `$A$` is an alignment or `-1`,
- [29+.#]{.pnum} `$W$` is a bit-field width or `-1`, and
- [29+.#]{.pnum} `$NUA$` is a boolean value.

Two data member descriptions are equal if each of their respective components are same types, same identifiers, and equal values.

::: note
The components of a data member description describe a data member such that

> - [29+.#]{.pnum} its type is specified using the type or type alias given by `$T$`,

Reviewer: We don't really do sub-paragraph numbering in notes.

Author: https://eel.is/c++draft/temp.res.general#note-4

I did check.

Reviewer: Huh. Ok.

[#]{.pnum} *Returns*:

* [#.#]{.pnum} If `r` is an unnamed entity other than a class that has a typedef name for linkage purposes ([dcl.typedef]{.sref}), then `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a class type `$C$`, then `true` when either the `$class-name$` of `$C$` is an identifier or `$C$` has a typedef name for linkage purposes. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a function, then `true` if the function is not a function template specialization, constructor, destructor, operator function, or conversion function. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a function template, then `true` if `r` does not represent a constructor template, operator function template, or conversion function template. Otherwise, `false`.

> * [#.#]{.pnum} Otherwise, if `r` represents a type alias, then `true` when the type alias is not a specialization of an alias template. Otherwise, `false`.

Reviewer: Could this be "... then `!has_template_arguments(r)`."?

Author: Great idea, done.

Author: Can do the same for the "variable" bullet; folded them both into one.

* [#.#]{.pnum} Otherwise, if `r` represents a variable, then `true` if `r` does not represent a variable template specialization. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a structured binding, enumerator, non-static data member, template, namespace, or namespace alias, then `true`. Otherwise, `false`.

> * [#.#]{.pnum} Otherwise, if `r` represents a direct base class relationship, then `true` if `has_identifier(type_of(r))`. Otherwise, `false`.

Reviewer: ```suggestion
* [#.#]{.pnum} Otherwise, if `r` represents a direct base class relationship, then `has_identifier(type_of(r))`.
```

Author: Done.

> * [#.#]{.pnum} Otherwise if `r` represents a data member description (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) ([class.mem.general]{.sref}), then `true` if `$N$` is not `-1`. Otherwise, `false`.

Reviewer: ```suggestion
* [#.#]{.pnum} Otherwise if `r` represents a data member description (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) ([class.mem.general]{.sref}), then `$N$ != -1`.
```
Although it's weird here that `$N$` is either a string or an integer?

Author: It is, but it's exposition-only bullshit. Maybe it's fine.

* [#.#]{.pnum} If `r` represents a literal operator or literal operator template, then the `$ud-suffix$` of the operator or operator template.
* [#.#]{.pnum} Otherwise, if `r` represents a class type, then either the typedef name for linkage purposes or the identifier introduced by the declaration of the represented type.
* [#.#]{.pnum} Otherwise, if `r` represents an entity, then the identifier introduced by the declaration of that entity.
* [#.#]{.pnum} Otherwise, if `r` represents a direct base class relationship, then `identifier_of(type_of(r))` or `u8identifier_of(type_of(r))`, respectively.

> * [#.#]{.pnum} Otherwise `r` represents a data member description (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) ([class.mem.general]{.sref}). A `string` or `u8string` respectively containing the identifier `$N$` encoded with `$E$`.

Reviewer: ```suggestion
* [#.#]{.pnum} Otherwise, `r` represents a data member description (`$T$`, `$N$`, `$A$`, `$W$`, `$NU$`) ([class.mem.general]{.sref}); a `string` or `u8string` respectively containing the identifier `$N$` encoded with `$E$`.
```

Author: Done.

* [#.#]{.pnum} Otherwise, if `r` represents a class type `$C$`, then `true` when either the `$class-name$` of `$C$` is an identifier or `$C$` has a typedef name for linkage purposes. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a function, then `true` if the function is not a function template specialization, constructor, destructor, operator function, or conversion function. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a function template, then `true` if `r` does not represent a constructor template, operator function template, or conversion function template. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents variable or a type alias, then `!has_template_arguments(r)`.
* [#.#]{.pnum} Otherwise, if `r` represents a structured binding, enumerator, non-static data member, template, namespace, or namespace alias, then `true`. Otherwise, `false`.
* [#.#]{.pnum} Otherwise, if `r` represents a direct base class relationship, then `has_identifier(type_of(r))`.
* [#.#]{.pnum} Otherwise if `r` represents a data member description (`$T$`, `$N$`, `$A$`, `$W$`, `$NUA$`) ([class.mem.general]{.sref}), then `$N$ != -1`.

> * [#.#]{.pnum} Otherwise, `false`.

Reviewer: Is this last bullet possible?

Author: Nope; good call. Updated.

consteval bool is_private(info r);
```

[#]{.pnum} *Returns*: `true` if `r` represents a class member or direct base class relationship that is public, protected, or private, respectively. Otherwise, `false`.

```cpp
consteval bool is_virtual(info r);
```

> [#]{.pnum} *Returns*: `true` if `r` represents either a virtual member function or a direct base class relationship with a virtual base class. Otherwise, `false`.

Reviewer: "direct base class relationship with a virtual base class" doesn't make sense to me. Should it be "direct base class relationship that is virtual"?

consteval bool is_base(info r);
```

> [#]{.pnum} *Returns*: `true` if `r` represents a class member, namespace member, non-static data member, static member, direct base class relationship, respectively. Otherwise, `false`.

Reviewer: Pre-existing, but this is missing an "or"

Author: Good catch, done.
